CRM/Contact/Form/Edit/Address.php

```diff
@@ -166,7 +166,10 @@ public static function buildQuickForm(&$form, $addressBlockCount = NULL, $sharin
 
       // Override the default profile links to add address form
       $profileLinks = CRM_Contact_BAO_Contact::getEntityRefCreateLinks('shared_address');
-      $form->addEntityRef("address[$blockId][master_contact_id]", ts('Share With'), array('create' => $profileLinks));
+      $form->addEntityRef("address[$blockId][master_contact_id]", ts('Share With'), ['create' => $profileLinks, 'api' => ['extra' => ['contact_type']]]);
```

Comment:
@samuelsov by sending back the contact type here as extra data, I was able to avoid the extra ajax callback from your original PR.